### PR TITLE
STATISTICS-27: Add truncated normal distribution.

### PR DESCRIPTION
**Pull request by mmalenic**
Truncated normal distribution implementation according to https://issues.apache.org/jira/browse/STATISTICS-27.

### CODE REVIEW

**Comment by aherbert** [comment]
Thanks for the contribution. The implementation and tests look good (all edge cases are covered).

The code is currently failing PMD checks. Most of these are for missing the `final` keyword on local variables. You can see the PMD errors by running:
```
mvn pmd:pmd
```
Then open the file in `target/site/pmd.html`

There is also an error for a confusing ternary. I think it is OK in this context but to avoid the issue what do you think to rewriting the constructor as:
```java
        if (lower == Double.NEGATIVE_INFINITY) {
            if (upper == Double.POSITIVE_INFINITY) {
                // No truncation
                this.mean = mean;
                variance = sd * sd;
            } else {
                // One-sided lower tail truncation
                double betaRatio = pdfBeta / cdfBeta;
                this.mean = mean - sd * betaRatio;
                variance = sd * sd * (1 - beta * betaRatio - betaRatio * betaRatio);
            }
        } else {
            if (upper == Double.POSITIVE_INFINITY) {
                // One-sided upper tail truncation
                double alphaRatio =  pdfAlpha / cdfDelta;
                this.mean = mean + sd * alphaRatio;
                variance = sd * sd * (1 + alpha * alphaRatio - alphaRatio * alphaRatio);
            } else {
                // Two-sided truncation
                this.mean = mean + pdfCdfDelta * parentSd;
                variance = sd * sd * (1 + alphaBetaDelta - pdfCdfDelta * pdfCdfDelta);
            }
        }
```

File: commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+     *
+     * @param mean mean for this distribution.
+     * @param sd standard deviation for this distribution.
+     * @param lower lower bound (inclusive) of the distribution, can be {@link Double#NEGATIVE_INFINITY}.
+     * @param upper upper bound (inclusive) of the distribution, can be {@link Double#POSITIVE_INFINITY}.
+     * @throws IllegalArgumentException if {@code sd <= 0} or if {@code upper <= lower}.
+     */
+    public TruncatedNormalDistribution(double mean, double sd, double lower, double upper) {
+        if (sd <= 0) {
+            throw new DistributionException(DistributionException.NEGATIVE, sd);
```
Comment (aherbert) [inline commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java:78]: I am thinking that we should have a `DistributionException.NOT_STRICTLY_POSITIVE` constant here. However the same message is used in NormalDistribution (and others) for zero so if you introduce this new constant then I can propagate it where appropriate through the rest of the code. Or you can leave it and I can do this after this PR is merged.

File: commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+     *
+     * Represents the true variance of the truncated normal distribution rather
+     * than the parent normal distribution variance.
+     */
+    @Override
+    public double getVariance() {
+        return variance;
+    }
+
+    /**
```
Comment (aherbert) [inline commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java:183]: In this case where nothing is added to the javadoc the rest of the codebase uses a single line `/** {@inheritDoc} */`

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/TruncatedNormalDistributionTest.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+                1.13584123966337e-07
+        );
+    }
+
+    /**
+     * Test mean and variance moments.
+     */
+    @Test
+    void testMoments() {
+        Assertions.assertEquals(mean, distribution.getMean(), TOLERANCE);
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/TruncatedNormalDistributionTest.java:264]: Can use `getTolerance()` here and avoid a class level constant

File: commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+ * Implementation of the truncated normal distribution.
+ *
+ * @see <a href="http://en.wikipedia.org/wiki/Truncated_normal_distribution">
+ * Truncated normal distribution (Wikipedia)</a>.
+ */
+public class TruncatedNormalDistribution extends AbstractContinuousDistribution {
+    /**
+     * Mean of parent normal distribution.
+     */
+    private final double parentMean;
```
Comment (aherbert) [inline commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java:30]: Single line javadoc can be left on one line: `/** Something short. */`

**Review by aherbert — CHANGES_REQUESTED** [review]
Thanks for the contribution.

The test class requires the testing of multiple instances of the same distribution. This case is not covered by the current abstract base class which assumes a single distribution is to be tested. The way you have implemented multiple cases for the distribution is fine. I will have a think on how the test suite can be reconfigured to better handle this situation.

For now I think there are just some minor issues to resolve and this can be merged.

**Comment by mmalenic** [comment]
Hi Alex, thanks for the feedback, I've implemented those changes.

I was thinking about the test class as well and while I was writing it I was attempting to use parameterized tests, however I wasn't able to come up with an elegant solution. Perhaps the abstract base class could somehow use parameterized tests?

In terms of adding distribution exceptions, how do you feel about adding a similar exception here: 
https://github.com/apache/commons-statistics/blob/1339b4ce6e1111217525bee893f1365b93a68acf/commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java#L66-L68
Which indicates that the condition is too large or equal to?

**Comment by aherbert** [comment]
I agree with the parameterised tests. This is a change that can be implemented later. We can discuss the approach on the mailing list.

I think the `TOO_LARGE` and `TOO_SMALL` apply when one of the arguments is fixed and the other is incorrect relative to it. The case here is for a range that is not specified correctly. So perhaps we should use an `INVALID_RANGE` exception where the upper bound is not above the lower bound:

`Lower bound %s is not below the upper bound %s`

This allows you to specify the range arguments in their declared order:

```java
if (upper <= lower) {
       throw new DistributionException(DistributionException.INVALID_RANGE, lower, upper);
}
```

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/TruncatedNormalDistributionTest.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+     * @param expectedMean expected mean.
+     * @param expectedVariance expected variance.
+     */
+    private void testAdditionalDistribution(
+            TruncatedNormalDistribution testDistribution,
+            double[] expectedPpfValues,
+            double[] expectedCdfValues,
+            double[] expectedPdfValues,
+            double expectedMean,
+            double expectedVariance
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/TruncatedNormalDistributionTest.java:98]: Remove the extra newline before the `) {`

File: commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/TruncatedNormalDistributionTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        testAdditionalDistribution(
+                new TruncatedNormalDistribution(12, 2.4, Double.NEGATIVE_INFINITY, 7.1),
+                new double[]{0, 0.00108276414971883, 0.00433032247708514, 0.0155754809421998, 0.0504271331622245,
+                             0.147106879016387, 0.387159643321778, 0.920668099879139, 1},
+                new double[]{Double.NEGATIVE_INFINITY, 2.20249292901062, 3.00511196424565, 3.80773099948069,
+                             4.61035003471573, 5.41296906995077, 6.21558810518581, 7.01820714042084, 7.1},
+                new double[]{0, 0.00194181137319567, 0.00719165311538403, 0.0238165586714952, 0.0705273999981105,
+                             0.186752027463317, 0.442182309739316, 0.936194292830215, 1.00423817618302},
+                6.21558810518581,
+                0.644197315721623
```
Comment (aherbert) [inline commons-statistics-distribution/src/test/java/org/apache/commons/statistics/distribution/TruncatedNormalDistributionTest.java:138]: Remove the extra new line before `);`

File: commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    /** A standard normal distribution used for calculations. */
+    private final NormalDistribution standardNormal;
+    /** Stored value of @{code standardNormal.cumulativeProbability((lower - mean) / sd)} for faster computations. */
+    private final double cdfAlpha;
+    /**
+     * Stored value of @{code standardNormal.cumulativeProbability((upper - mean) / sd) - cdfAlpha}
+     * for faster computations.
+     */
+    private final double cdfDelta;
+
```
Comment (aherbert) [inline commons-statistics-distribution/src/main/java/org/apache/commons/statistics/distribution/TruncatedNormalDistribution.java:49]: Remove the double line break.

**Review by aherbert — CHANGES_REQUESTED** [review]
Thanks for the changes. I have noted some minor formatting nits. You can update the TOO_LARGE exception to some kind of range exception.

**Comment by mmalenic** [comment]
Sounds good, I've just done those changes, let me know if anything else should be fixed.

**Comment by aherbert** [comment]
All looks good. Can you squash all commits and prefix the commit message with `[STATISTICS-27]`.

**Review by brentworden — APPROVED** [review]
